From 76322f2c2ee19ff84ea4a7297ce49e0c7b97567e Mon Sep 17 00:00:00 2001
From: Balaji Eswaramurthy <beswaramurthy@maxlinear.com>
Date: Fri, 25 Oct 2024 12:55:57 +0800
Subject: [PATCH] WLANRTSYS-85881 Fix GTK rekeying EAPOL M1 length calculation

Issue: Length calculation for GTK rekeying EAPOL M1 is wrong.

Fix: mxl_ieee80211w_kde_len() calculates length for both IGTK KDE and
BIGTK KDE. Skip mxl_ieee80211w_kde_len() call in BIGTK KDE length
calculation.

FIXUP to WLANRTSYS-62845 MLO changes: MLO Auth Assoc, WPA3 and eapol
changes.
---
 src/ap/mxl_wpa_auth.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/ap/mxl_wpa_auth.c b/src/ap/mxl_wpa_auth.c
index 8af7537b3..8398378eb 100644
--- a/src/ap/mxl_wpa_auth.c
+++ b/src/ap/mxl_wpa_auth.c
@@ -565,24 +565,28 @@ void mxl_ml_wpa_ptk_group_rekeynegotiating(struct wpa_state_machine *sm)
 		/* MLO Link1 GTK KDE length */
 		kde_len = (2 + 4 + 1 + PN_LENGTH + gsm->GTK_len);
 
-		/* MLO Link1 IGTK KDE length */
+		/* MLO Link1 IGTK KDE length.
+		 * mxl_ieee80211w_kde_len() calculates length for IGTK KDE and BIGTK KDE.
+		 */
 		kde_len += mxl_ieee80211w_kde_len(sm) + 1; /* Link info */
 
 		/* MLO Link1 BIGTK KDE length */
 		if (sm->wpa_auth->conf.beacon_prot)
-			kde_len += mxl_ieee80211w_kde_len(sm) + mxl_ocv_oci_len(sm) + 1; /* Link info */
+			kde_len += mxl_ocv_oci_len(sm) + 1; /* Link info */
 
 		if (sibling_wpa_auth && !sm->mxl_sm.is_single_link_ml_sta) {
 			/* MLO Link2 GTK KDE length */
 			gsm1 = sibling_wpa_auth->group;
 			kde_len += (2 + 4 + 1 + PN_LENGTH + gsm1->GTK_len);
 
-			/* MLO Link2 IGTK KDE length */
+			/* MLO Link2 IGTK KDE length.
+			 * mxl_ieee80211w_kde_len() calculates length for IGTK KDE and BIGTK KDE.
+			 */
 			kde_len += mxl_ieee80211w_kde_len(sm) + 1; /* Link info */
 
 			/* MLO Link2 BIGTK KDE length */
 			if (sibling_wpa_auth->conf.beacon_prot)
-				kde_len += mxl_ieee80211w_kde_len(sm) + mxl_ocv_oci_len(sm) + 1; /* Link info */
+				kde_len += mxl_ocv_oci_len(sm) + 1; /* Link info */
 		}
 
 		kde_buf = os_malloc(kde_len);
-- 
2.43.0

